src/Agda2Hs/Compile/RuntimeCheckUtils.hs

```diff
@@ -0,0 +1,252 @@
+    -- Partition out arguments that are erased and at top level (those we will attempt to check)
+    (topLevelErased, call) = partition (checkTopLevelErased . fst) $ zip doms telsUpTo
+
+-- Check a single type for runtime checkability.
+-- Accumulates on name indices for `go` function and `a` argument.
+-- Takes domain of type and telescope up to that point for context.
+-- If checkable, returns lhs and rhs at that point plus declarations from checks below.
+checkRtc'' :: NameIndices -> (Dom (ArgName, Type), Telescope) -> C (NameIndices, Maybe (Hs.Pat (), Hs.Exp (), [Hs.Decl ()]))
+checkRtc'' (goIdx, argIdx) (d, tUpTo) = do
+  ((theirGoIdx, theirArgIdx), chk) <- addContext tUpTo (checkRtc' (goIdx, ourArgIdx) tAt)
+  case chk of
+    NoneErased' -> return ((theirGoIdx, theirArgIdx), Just (ourLhs, hsVar arg, []))
+    Uncheckable' -> return ((theirGoIdx, theirArgIdx), Nothing)
+    Checkable' {..} -> do
+      let go = "go" ++ show theirGoIdx
+          conflicts = tAtNames `intersect` [go, arg, up]
+      unless (null conflicts) $ errorNameConflict $ head conflicts
+      let rhs = hsVar up `eApp` theirRhs
+          maybeRtc = createRtc (hsName go) (hsPat up : theirLhs) theirChks rhs $ binds theirDecls
+          (ourGoIdx, ourRhs, ourRtc) =
+            if null theirChks
+              then -- inline if nothing to check at this level (consumes no `goIdx`)
+                (theirGoIdx, Hs.Lambda () theirLhs $ hsVar arg `eApp` theirRhs, theirDecls)
+              else (succ theirGoIdx, hsVar go `eApp` [hsVar arg], [maybeRtc])
+      return ((ourGoIdx, theirArgIdx), Just (ourLhs, ourRhs, ourRtc))
+  where
+    tAt = telify d
+    tAtNames = map (fst . unDom) $ telToList tAt
+    name = fst $ unDom d
+    -- Use arg name if available, otherwise insert one (consumes one on `argIdx`)
+    (arg, ourArgIdx) = if name == "_" then ("a" ++ show argIdx, succ argIdx) else (name, argIdx)
+    ourLhs = hsPat arg
+    up = "up"
+
+---- Small single-mention convenience functions
+
+-- Gather telescope from domain
+telify :: Dom (a, Type) -> Telescope
+telify = theTel . telView' . snd . unDom
+
+-- Check a type is erased and at top level; in this case, it should be checked.
+checkTopLevelErased :: Dom (a, Type) -> Bool
+checkTopLevelErased dom = case getQuantity dom of
+  Quantity0 _ -> null $ telToList $ telify dom
+  _ -> False
+
+-- Create binds from declarations except when empty
+binds :: [Hs.Decl ()] -> Maybe (Hs.Binds ())
+binds [] = Nothing
+binds decls = Just $ Hs.BDecls () decls
+
+-- Turn a type into its Dec version
+decify :: Type -> C Type
+decify t = do
+  dec <- resolveStringName "Haskell.Extra.Dec.Dec"
+  level <- liftTCM newLevelMeta
+  let vArg = defaultArg
+      hArg = setHiding Hidden . vArg
+  return $ t {unEl = Def dec $ map Apply [hArg $ Level level, vArg $ unEl t]}
+
+-- Failably find instances for decidable terms
+findDecInstances :: Type -> C (Maybe Term)
+findDecInstances t =
+  liftTCM $
```

Comment:
Typically I would put both `liftTCM` and `do` on the same line as `findDecInstances t =`, saving on indentation and vertical space.